File: nmos/registry/paging.py
# Copyright (C) 2025-2026 Alain Bouchard
# SPDX-License-Identifier: Apache-2.0

"""Query API pagination.

Implements the ``paged`` trait of ``QueryAPI.raml:25-63`` and the worked
examples of ``APIs - Query Parameters.md:34-369``. The examples are the real
specification here — the prose leaves several boundary cases implicit and only
the nine worked cases pin them down — so this module is written to reproduce
all nine exactly, and ``tests/test_paging.py`` asserts each one.

The model
---------
A page is a window over the collection ordered by a registry-assigned cursor
(creation or update time, selected by ``paging.order``). Two rules fix
everything else:

* ``paging.since`` is **non-inclusive**, ``paging.until`` is **inclusive**
  (``QueryAPI.raml:29,33``). The window is therefore ``(since, until]``.
* The payload is always returned **most recent first**
  (``APIs - Query Parameters.md:90``, ``QueryAPI.raml:40``).

Which end of the window the page is taken from depends on which cursor the
client supplied, and that is the part worth being explicit about:

* ``since`` given — page forwards from the bottom of the window. The client is
  walking towards newer records, so the page is the *oldest* ``limit`` records
  above ``since``, and ``X-Paging-Until`` reports where it got to.
* ``since`` absent — page backwards from the top. The client is looking at the
  newest records, so the page is the *newest* ``limit`` records at or below
  ``until``, and ``X-Paging-Since`` reports how far down it reached.

That asymmetry is why ``since`` "takes precedence where a resulting data set is
constrained by the server's value of limit" (``:30``, Example 5): with both
cursors supplied and more than ``limit`` records between them, the page is
anchored to ``since`` and the far end is reported back narrowed.

Cursor reporting is not just the min/max of what was returned. ``X-Paging-Until``
with no cursors supplied is the newest cursor in the **unfiltered** collection,
not of the filtered page — Edge Cases 3 and 4 both return a page whose newest
record is older than the reported ``until``, or no records at all, and still
report the collection's own maximum. That is what makes the ``next`` cursor
usable as a "watch for changes from now on" bookmark even when the current
filter matches nothing.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Iterable, Sequence
from urllib.parse import quote, urlencode

from nmos.registry.types import RegisteredResource, TaiCursor

# ``paging.order`` values, ``QueryAPI.raml:39-43``. Default is ``update``.
ORDER_CREATE = "create"
ORDER_UPDATE = "update"
VALID_ORDERS = frozenset({ORDER_CREATE, ORDER_UPDATE})

# Query-string parameters this module owns. Everything else on the query
# string is a basic-query filter -- see ``query_filter.py``.
PARAM_SINCE = "paging.since"
PARAM_UNTIL = "paging.until"
PARAM_LIMIT = "paging.limit"
PARAM_ORDER = "paging.order"
PAGING_PARAMS = frozenset({PARAM_SINCE, PARAM_UNTIL, PARAM_LIMIT, PARAM_ORDER})


class PagingError(Exception):
    """A paging parameter was malformed. The caller answers 400."""


@dataclass(frozen=True)
class PagingRequest:
    """The paging parameters of one request, parsed and bounds-checked."""

    since: TaiCursor | None
    until: TaiCursor | None
    limit: int
    order: str

    @property
    def order_by_created(self) -> bool:
        return self.order == ORDER_CREATE

    def cursor_of(self, resource: RegisteredResource) -> TaiCursor:
        """The cursor this request pages on."""
        return resource.created if self.order_by_created else resource.updated


@dataclass(frozen=True)
class Page:
    """A rendered page: the records plus the cursors describing the window."""

    resources: list[RegisteredResource]
    since: TaiCursor
    until: TaiCursor
    limit: int


def parse_paging(
    params: dict[str, str], *, default_limit: int, max_limit: int,
) -> PagingRequest:
    """Parse the four ``paging.*`` query parameters.

    Args:
        params: The request's query string as a flat mapping.
        default_limit: Server default page size when the client asks for none.
        max_limit: Ceiling the server will honour. A larger request is clamped
            rather than refused — ``:137``: "If the client had requested a page
            size which the server was unable to honour, the actual page size
            used would be returned in X-Paging-Limit".

    Raises:
        PagingError: A cursor does not match ``^[0-9]+:[0-9]+$``, the limit is
            not a positive integer, or the order is not ``create``/``update``.
    """
    since = _cursor_param(params, PARAM_SINCE)
    until = _cursor_param(params, PARAM_UNTIL)

    limit = default_limit
    raw_limit = params.get(PARAM_LIMIT)
    if raw_limit is not None:
        # The RAML types this as an integer; a non-integer is malformed input
        # rather than something to silently round or ignore.
        if not raw_limit.lstrip("-").isdigit():
            raise PagingError(f"{PARAM_LIMIT} must be an integer, got {raw_limit!r}")
        limit = int(raw_limit)
        # Zero is a legitimate limit, not an error: it asks "give me the paging
        # cursors for this window but none of the records", which is how a
        # client probes a collection's bounds cheaply. AMWA IS-04-02
        # ``test_21_4`` exercises it directly and expects a 200 with an empty
        # body and ``X-Paging-Limit: 0``. Only a negative limit is malformed.
        if limit < 0:
            raise PagingError(f"{PARAM_LIMIT} must not be negative, got {limit}")
        limit = min(limit, max_limit)

    order = params.get(PARAM_ORDER, ORDER_UPDATE)
    if order not in VALID_ORDERS:
        raise PagingError(
            f"{PARAM_ORDER} must be one of {sorted(VALID_ORDERS)}, got {order!r}",
        )

    # An inverted window describes no possible page: `since` is the exclusive
    # lower bound and `until` the inclusive upper one, so since > until asks
    # for records both newer and older than each other. That is a malformed
    # request rather than a request with no matches, and is answered 400.
    #
    # since == until is NOT an error -- it is the legitimate "has anything
    # arrived since this cursor?" poll, and must return an empty page with
    # paging headers so the client can keep re-issuing the same cursor
    # (APIs - Query Parameters.md:313).
    if since is not None and until is not None and since > until:
        raise PagingError(
            f"{PARAM_SINCE} ({since}) must not be later than "
            f"{PARAM_UNTIL} ({until})",
        )

    return PagingRequest(since=since, until=until, limit=limit, order=order)


def _cursor_param(params: dict[str, str], name: str) -> TaiCursor | None:
    raw = params.get(name)
    if raw is None:
        return None
    cursor = TaiCursor.parse(raw)
    if cursor is None:
        raise PagingError(
            f"{name} must match '<seconds>:<nanoseconds>', got {raw!r}",
        )
    return cursor


def apply_paging(
    matched: Iterable[RegisteredResource],
    collection: Iterable[RegisteredResource],
    request: PagingRequest,
) -> Page:
    """Select one page from a filtered collection.

    Args:
        matched: The resources that passed the basic-query filters. ``:26``
            requires filtering to happen before paging, so this is already the
            filtered set.
        collection: Every extant resource of this type, unfiltered. Needed
            only to compute the default upper cursor — see the module
            docstring on Edge Cases 3 and 4.
        request: The parsed paging parameters.

    Returns:
        The page, most recent first, with the cursors to report in
        ``X-Paging-Since`` / ``X-Paging-Until``.
    """
    key = request.cursor_of
    # Sorted on (cursor, id), not on the cursor alone. The cursor is normally
    # unique within a type -- ``APIs - Query Parameters.md:17`` asks for that,
    # and the store enforces it by falling forward a nanosecond on collision --
    # but "normally" is not "always", and Python's sort is stable, so a tie
    # would otherwise be broken by dictionary insertion order.
    #
    # That is invisible in a single registry and wrong in a cluster: two members
    # that received the same two resources in a different order would return
    # the same page with its contents in a different order, and a client paging
    # across members would see records repeat or vanish. The id is arbitrary but
    # identical everywhere, which is all a tie-break has to be.
    ascending = sorted(matched, key=lambda resource: (key(resource), resource.id))

    if request.since is not None:
        # Paging forwards: the oldest `limit` records strictly above `since`,
        # bounded above by `until` when the client supplied one.
        window = [
            resource for resource in ascending
            if key(resource) > request.since
            and (request.until is None or key(resource) <= request.until)
        ]
        page = window[: request.limit]

        # `since` is echoed exactly as requested. `until` reports the top of
        # the window that was actually served, and the distinction that
        # matters is whether the limit *truncated* the window:
        #
        # * Truncated (more records matched than fit) -- narrow to the newest
        #   record served, so the client's `next` cursor resumes exactly where
        #   this page stopped. This is Example 5's "since takes precedence":
        #   X-Paging-Until comes back lower than requested.
        # * Not truncated -- the whole window was served, so the requested
        #   ceiling still describes it and must be echoed unchanged. Narrowing
        #   here would be wrong: it would move the client's `next` cursor
        #   backwards to the last *matching* record, so a later record that
        #   did not match the filter would be replayed on the next page.
        #   A filter selecting discontiguous records makes this visible, which
        #   is what AMWA IS-04-02 ``test_21_5`` checks.
        #
        # With nothing served at all, the window collapses onto the bound(s)
        # supplied: the requested ceiling if there was one, otherwise `since`
        # itself (Edge Case 2, where until == since makes the client re-issue
        # the identical cursor rather than skipping past records that have not
        # arrived yet).
        truncated = len(window) > request.limit
        if truncated and page:
            report_until = key(page[-1])
        elif request.until is not None:
            report_until = request.until
        elif page:
            report_until = key(page[-1])
        else:
            report_until = request.since

        return Page(
            resources=list(reversed(page)),
            since=request.since,
            until=report_until,
            limit=request.limit,
        )

    # Paging backwards: the newest `limit` records at or below `until`.
    # Absent an explicit `until`, the ceiling is the newest cursor in the
    # UNFILTERED collection.
    ceiling = request.until if request.until is not None else _max_cursor(
        collection, key,
    )
    window = [resource for resource in ascending if key(resource) <= ceiling]
    # ``window[-0:]`` is the whole list, not an empty one, so a zero limit has
    # to be handled before the slice rather than falling out of it.
    if request.limit == 0:
        page = []
    else:
        page = window[-request.limit:] if request.limit <= len(window) else window

    # `since` is the exclusive lower bound that reproduces exactly this page:
    # the cursor of the record immediately below it. When the page did not
    # fill, the window reaches the bottom of the collection and the bound is
    # 0:0 (Edge Cases 1, 3 and 4).
    #
    # A zero limit served nothing, so there is no "record below the page" --
    # the window has no extent and both bounds sit on the ceiling.
    if request.limit == 0:
        report_since = ceiling
    elif len(window) > request.limit:
        report_since = key(window[-request.limit - 1])
    else:
        report_since = TaiCursor.min()

    return Page(
        resources=list(reversed(page)),
        since=report_since,
        until=ceiling,
        limit=request.limit,
    )


def _max_cursor(
    collection: Iterable[RegisteredResource],
    key: Callable[[RegisteredResource], TaiCursor],
) -> TaiCursor:
    """Newest cursor in the collection, or 0:0 when it is empty."""
    highest = TaiCursor.min()
    for resource in collection:
        cursor = key(resource)
        if cursor > highest:
            highest = cursor
    return highest


def paging_headers(
    page: Page, base_url: str, filters: Sequence[tuple[str, str]], order: str,
) -> dict[str, str]:
    """Build the ``X-Paging-*`` and ``Link`` response headers.

    ``:13`` makes ``X-Paging-Limit`` mandatory whenever paging is in use —
    it is how a client detects that the API pages at all — so all three
    ``X-Paging-*`` headers are always emitted.

    ``:32`` says servers SHOULD return ``prev`` and ``next`` and MAY return
    ``first`` and ``last``; all four are provided. Their construction follows
    nmos-cpp: ``next`` pages upward from the window's top cursor, ``prev``
    downward from its bottom, ``first`` is ``paging.since=0:0`` (``:100``),
    and ``last`` carries no paging cursors at all, which is by definition the
    newest page (``:99``).

    Args:
        page: The page just produced.
        base_url: Absolute URL of this collection, used verbatim as the link
            target so the links stay valid behind a proxy.
        filters: The non-paging query parameters, preserved on every link —
            Edge Case 3 shows ``label=My%20Node`` carried through.
        order: The effective ``paging.order``, echoed only when it is not the
            default so common URLs stay short.
    """
    def encode(query: list[tuple[str, str]]) -> str:
        """Percent-encode a query string, leaving ``:`` literal.

        The colon matters. A client comparing the ``prev`` link against the
        ``X-Paging-Since`` header it was given does a plain string match, so
        ``paging.until=1441716120:318744030`` and
        ``paging.until=1441716120%3A318744030`` are not interchangeable even
        though they are the same URL. Every cursor in the specification's
        worked examples is written with a literal colon
        (``APIs - Query Parameters.md:65``), and RFC 3986 permits ``:`` in a
        query, so that is what is emitted.

        ``quote`` rather than the default ``quote_plus``: a space becomes
        ``%20``, matching the specification's ``?label=My%20Node`` (Edge Case
        3, ``:330``) rather than ``+``. Filter values are otherwise fully
        encoded -- an ``&`` inside a value still becomes ``%26``, so a label
        of ``foo&bar`` round-trips instead of splitting the query.
        """
        return urlencode(query, safe=":", quote_via=quote)

    def link(cursor_param: str, cursor: TaiCursor) -> str:
        query: list[tuple[str, str]] = list(filters)
        query.append((cursor_param, str(cursor)))
        query.append((PARAM_LIMIT, str(page.limit)))
        if order != ORDER_UPDATE:
            query.append((PARAM_ORDER, order))
        return f"<{base_url}?{encode(query)}>"

    def bare_link() -> str:
        query = list(filters)
        query.append((PARAM_LIMIT, str(page.limit)))
        if order != ORDER_UPDATE:
            query.append((PARAM_ORDER, order))
        return f"<{base_url}?{encode(query)}>"

    links = [
        f'{link(PARAM_SINCE, page.until)}; rel="next"',
        f'{link(PARAM_UNTIL, page.since)}; rel="prev"',
        f'{link(PARAM_SINCE, TaiCursor.min())}; rel="first"',
        f'{bare_link()}; rel="last"',
    ]

    return {
        "Link": ", ".join(links),
        "X-Paging-Limit": str(page.limit),
        "X-Paging-Since": str(page.since),
        "X-Paging-Until": str(page.until),
    }

File: rust/crates/nmos-registry-core/tests/paging.rs
// Copyright (C) 2025-2026 Alain Bouchard
// SPDX-License-Identifier: Apache-2.0

//! Pagination, ported from `nmos/registry/tests/test_paging.py`.
//!
//! The nine worked examples of `APIs - Query Parameters.md:34-369` **are** the
//! specification for this area: the prose leaves several boundary cases
//! implicit and only the examples pin them down. All nine are here, each
//! naming the lines it comes from.
//!
//! Where the AMWA mock registry differs, the test says so. Those are not
//! gratuitous comparisons -- they are the cases most likely to be "fixed" back
//! to the wrong behaviour by someone who checked the mock rather than the
//! document.

#![allow(
    clippy::unwrap_used,
    clippy::expect_used,
    clippy::panic,
    clippy::indexing_slicing
)]

use nmos_registry_core::body::Body;
use nmos_registry_core::cursor::TaiCursor;
use nmos_registry_core::paging::{Page, apply_paging, paging_headers, parse_paging};
use nmos_registry_core::resource::{Order, RegisteredResource};
use nmos_registry_core::resource_type::ResourceType;
use serde_json::json;

/// The examples use a server whose default paging limit is 10 (`:52`).
const DEFAULT_LIMIT: usize = 10;
const MAX_LIMIT: usize = 100;
const BASE_URL: &str = "http://api.example.com/x-nmos/query/v1.1/nodes/";

/// One sample record, identified only by its cursor.
fn resource(nanoseconds: u32) -> RegisteredResource {
    let cursor = TaiCursor::new(0, nanoseconds);
    RegisteredResource::new(
        ResourceType::Node,
        format!("00000000-0000-1000-8000-{nanoseconds:012}"),
        Body::from_value(json!({"id": format!("node-{nanoseconds}")})),
        cursor.to_string(),
        cursor,
        cursor,
        None,
    )
}

/// `[0:1, 0:2, ... 0:20]` -- the sample data set of `:40-46`.
fn sample() -> Vec<RegisteredResource> {
    (1..=20).map(resource).collect()
}

fn refs(data: &[RegisteredResource]) -> Vec<&RegisteredResource> {
    data.iter().collect()
}

fn lookup(pairs: &[(&str, &str)]) -> impl Fn(&str) -> Option<String> + use<> {
    let owned: Vec<(String, String)> = pairs
        .iter()
        .map(|(k, v)| ((*k).to_owned(), (*v).to_owned()))
        .collect();
    move |name: &str| {
        owned
            .iter()
            .find(|(k, _)| k == name)
            .map(|(_, v)| v.clone())
    }
}

/// A page from `matched`, with the collection's newest cursor for the default
/// upper bound.
fn page_for<'a>(
    matched: &[&'a RegisteredResource],
    query: &[(&str, &str)],
    collection: Option<&[&RegisteredResource]>,
) -> Page<'a> {
    let request = parse_paging(lookup(query), DEFAULT_LIMIT, MAX_LIMIT).expect("valid paging");
    let source = collection.unwrap_or(matched);
    let collection_max = source.last().map(|r| request.cursor_of(r));
    apply_paging(matched, collection_max, &request)
}

fn cursors(page: &Page<'_>) -> Vec<String> {
    page.resources
        .iter()
        .map(|r| r.updated.to_string())
        .collect()
}

fn headers(page: &Page<'_>, filters: &[(&str, &str)], order: Order) -> Vec<(String, String)> {
    let owned: Vec<(String, String)> = filters
        .iter()
        .map(|(k, v)| ((*k).to_owned(), (*v).to_owned()))
        .collect();
    paging_headers(page.window(), BASE_URL, &owned, order)
}

fn header(set: &[(String, String)], name: &str) -> String {
    set.iter()
        .find(|(k, _)| k == name)
        .map(|(_, v)| v.clone())
        .unwrap_or_else(|| panic!("no {name} header"))
}

// ---------------------------------------------------------------------------
// The five numbered examples
// ---------------------------------------------------------------------------

#[test]
fn example_1_the_initial_request_returns_the_newest_page_descending() {
    // `:50-99` -- "The data set returned when no paging.since or paging.until
    // parameters are specified MUST be from the most recently updated (or
    // created) resources in the collection, returned in descending order."
    //
    // The AMWA mock returns the OLDEST ten, ascending.
    let data = sample();
    let page = page_for(&refs(&data), &[], None);

    assert_eq!(
        cursors(&page),
        [
            "0:20", "0:19", "0:18", "0:17", "0:16", "0:15", "0:14", "0:13", "0:12", "0:11",
        ],
    );
    assert_eq!(page.since.to_string(), "0:10");
    assert_eq!(page.until.to_string(), "0:20");
    assert_eq!(page.limit, 10);
}

#[test]
fn example_2_a_custom_limit_narrows_the_page_and_the_lower_cursor() {
    // `:102-137`.
    let data = sample();
    let page = page_for(&refs(&data), &[("paging.limit", "5")], None);

    assert_eq!(cursors(&page), ["0:20", "0:19", "0:18", "0:17", "0:16"]);
    assert_eq!(page.since.to_string(), "0:15");
    assert_eq!(page.until.to_string(), "0:20");
    assert_eq!(page.limit, 5);
}

#[test]
fn example_3_since_is_non_inclusive_and_pages_forwards() {
    // `:139-173`. The window opens at 0:5, not 0:4.
    let data = sample();
    let page = page_for(&refs(&data), &[("paging.since", "0:4")], None);

    assert_eq!(
        cursors(&page),
        [
            "0:14", "0:13", "0:12", "0:11", "0:10", "0:9", "0:8", "0:7", "0:6", "0:5"
        ],
    );
    assert_eq!(
        page.since.to_string(),
        "0:4",
        "since is echoed as requested"
    );
    assert_eq!(page.until.to_string(), "0:14");
}

#[test]
fn example_4_until_is_inclusive_and_pages_backwards() {
    // `:175-209`. 0:16 itself is returned.
    let data = sample();
    let page = page_for(&refs(&data), &[("paging.until", "0:16")], None);

    assert_eq!(
        cursors(&page),
        [
            "0:16", "0:15", "0:14", "0:13", "0:12", "0:11", "0:10", "0:9", "0:8", "0:7"
        ],
    );
    assert_eq!(page.since.to_string(), "0:6");
    assert_eq!(page.until.to_string(), "0:16");
}

#[test]
fn example_5_since_takes_precedence_and_narrows_the_reported_until() {
    // `:211-249` -- "Whilst both since and until are specified, as this server
    // example has a default paging limit of 10, the since parameter takes
    // precedence. As a result of this the value of X-Paging-Until is lower
    // than requested".
    let data = sample();
    let page = page_for(
        &refs(&data),
        &[("paging.since", "0:4"), ("paging.until", "0:16")],
        None,
    );

    assert_eq!(
        cursors(&page),
        [
            "0:14", "0:13", "0:12", "0:11", "0:10", "0:9", "0:8", "0:7", "0:6", "0:5"
        ],
    );
    assert_eq!(page.since.to_string(), "0:4");
    assert_eq!(
        page.until.to_string(),
        "0:14",
        "the requested ceiling of 0:16 was not narrowed to what was served",
    );
}

// ---------------------------------------------------------------------------
// The four edge cases
// ---------------------------------------------------------------------------

#[test]
fn edge_1_a_window_entirely_below_the_data_is_empty() {
    // `:257-282` -- `?paging.until=0:20` with only 0:21 and 0:22 stored.
    let data = vec![resource(21), resource(22)];
    let page = page_for(&refs(&data), &[("paging.until", "0:20")], None);

    assert!(cursors(&page).is_empty());
    assert_eq!(page.since.to_string(), "0:0");
    assert_eq!(page.until.to_string(), "0:20");
}

#[test]
fn edge_2_at_the_end_of_the_data_both_cursors_report_the_same_instant() {
    // `:284-313` -- `?paging.since=0:20` with only 0:19 and 0:20 stored.
    //
    // The `next` link then repeats the identical request, which is deliberate:
    // "the client is expected to re-perform the same request ... If the client
    // were to increment the value of since requested it would be in danger of
    // moving ahead of the current time and missing records."
    let data = vec![resource(19), resource(20)];
    let page = page_for(&refs(&data), &[("paging.since", "0:20")], None);

    assert!(cursors(&page).is_empty());
    assert_eq!(page.since.to_string(), "0:20");
    assert_eq!(page.until.to_string(), "0:20");
}

#[test]
fn edge_3_a_filter_matching_one_record_still_reports_the_collections_maximum() {
    // `:315-342`. The filtered record is at 0:15; the newest in the whole
    // collection is 0:20. `X-Paging-Until` reports 0:20 -- the collection's
    // maximum, not the page's -- so the `next` cursor remains a valid
    // "everything from now on" bookmark.
    let all = sample();
    let matched = vec![resource(15)];
    let page = page_for(&refs(&matched), &[], Some(&refs(&all)));

    assert_eq!(cursors(&page), ["0:15"]);
    assert_eq!(page.since.to_string(), "0:0");
    assert_eq!(
        page.until.to_string(),
        "0:20",
        "the page's own newest cursor was reported instead of the collection's",
    );
}

#[test]
fn edge_4_a_filter_matching_nothing_still_reports_the_collections_cursors() {
    // `:344-369`. An empty payload, but the headers still describe the window.
    let all = sample();
    let page = page_for(&[], &[], Some(&refs(&all)));

    assert!(cursors(&page).is_empty());
    assert_eq!(page.since.to_string(), "0:0");
    assert_eq!(page.until.to_string(), "0:20");
}

// ---------------------------------------------------------------------------
// Headers and links
// ---------------------------------------------------------------------------

#[test]
fn example_1_produces_the_documented_header_set() {
    // `:62-69`.
    let data = sample();
    let page = page_for(&refs(&data), &[], None);
    let set = headers(&page, &[], Order::Updated);

    assert_eq!(header(&set, "X-Paging-Limit"), "10");
    assert_eq!(header(&set, "X-Paging-Since"), "0:10");
    assert_eq!(header(&set, "X-Paging-Until"), "0:20");

    let link = header(&set, "Link");
    assert!(
        link.contains(&format!(
            "<{BASE_URL}?paging.since=0:20&paging.limit=10>; rel=\"next\""
        )),
        "{link}",
    );
    assert!(
        link.contains(&format!(
            "<{BASE_URL}?paging.until=0:10&paging.limit=10>; rel=\"prev\""
        )),
        "{link}",
    );
}

#[test]
fn the_links_are_string_comparable_against_the_paging_headers() {
    // A client -- and the AMWA suite -- matches the `prev` link against the
    // `X-Paging-Since` header it was handed, as plain text. A percent-encoded
    // colon would not compare equal even though the URL is equivalent.
    let all = vec![resource(20)];
    let matched = vec![resource(15)];
    let page = page_for(&refs(&matched), &[], Some(&refs(&all)));
    let set = headers(&page, &[], Order::Updated);
    let link = header(&set, "Link");

    assert!(
        !link.contains("%3A"),
        "a cursor colon was percent-encoded: {link}"
    );
    assert!(link.contains(&format!("paging.until={}", header(&set, "X-Paging-Since"))));
    assert!(link.contains(&format!("paging.since={}", header(&set, "X-Paging-Until"))));
}

#[test]
fn first_is_the_zero_cursor_and_last_carries_none() {
    // `:98-100`.
    let data = sample();
    let page = page_for(&refs(&data), &[], None);
    let link = header(&headers(&page, &[], Order::Updated), "Link");

    assert!(link.contains(&format!(
        "<{BASE_URL}?paging.since=0:0&paging.limit=10>; rel=\"first\""
    )));
    assert!(link.contains(&format!("<{BASE_URL}?paging.limit=10>; rel=\"last\"")));
}

#[test]
fn filters_are_preserved_on_every_link_with_percent_twenty() {
    // Edge Case 3's links carry `label=My%20Node` through -- `%20`, not `+`.
    let all = vec![resource(20)];
    let matched = vec![resource(15)];
    let page = page_for(&refs(&matched), &[], Some(&refs(&all)));
    let link = header(
        &headers(&page, &[("label", "My Node")], Order::Updated),
        "Link",
    );

    assert!(link.contains("label=My%20Node"), "{link}");
    assert_eq!(
        link.matches("label=My%20Node").count(),
        4,
        "the filter must ride on all four links",
    );
}

#[test]
fn an_ampersand_in_a_filter_value_is_still_encoded() {
    // Leaving `:` literal must not leave `&` literal too. AMWA `test_21_8`
    // queries `?label=foo%26bar`; an unencoded `&` would split the query
    // string and change its meaning.
    let all = vec![resource(20)];
    let matched = vec![resource(15)];
    let page = page_for(&refs(&matched), &[], Some(&refs(&all)));
    let link = header(
        &headers(&page, &[("label", "foo&bar")], Order::Updated),
        "Link",
    );

    assert!(link.contains("label=foo%26bar"), "{link}");
}

#[test]
fn a_non_default_order_is_echoed_on_the_links() {
    // A `create`-ordered page must produce `create`-ordered links, or the
    // client's next request silently switches ordering.
    let data = sample();
    let page = page_for(&refs(&data), &[("paging.order", "create")], None);
    let link = header(&headers(&page, &[], Order::Created), "Link");

    assert!(link.contains("paging.order=create"), "{link}");
    assert_eq!(link.matches("paging.order=create").count(), 4);
}

#[test]
fn the_default_order_is_not_echoed() {
    // So common URLs stay short.
    let data = sample();
    let page = page_for(&refs(&data), &[], None);
    let link = header(&headers(&page, &[], Order::Updated), "Link");
    assert!(!link.contains("paging.order"), "{link}");
}

// ---------------------------------------------------------------------------
// Ordering
// ---------------------------------------------------------------------------

#[test]
fn order_by_create_pages_on_the_creation_cursor() {
    // Built so the two orderings disagree: the record created first is
    // updated last. A page that ignored `paging.order` would return them the
    // same way round either way.
    let mut first = resource(1);
    first.updated = TaiCursor::new(0, 99);
    let mut second = resource(2);
    second.updated = TaiCursor::new(0, 3);

    // Ascending by *update*: second (0:3) then first (0:99).
    let by_update_data = vec![&second, &first];
    let update_request = parse_paging(lookup(&[]), DEFAULT_LIMIT, MAX_LIMIT).unwrap();
    let by_update = apply_paging(&by_update_data, Some(first.updated), &update_request);
    assert_eq!(
        by_update
            .resources
            .iter()
            .map(|r| r.id.as_str())
            .collect::<Vec<_>>(),
        [first.id.as_str(), second.id.as_str()],
        "update order should put the most recently updated first",
    );

    // Ascending by *create*: first (0:1) then second (0:2).
    let by_create_data = vec![&first, &second];
    let create_request = parse_paging(
        lookup(&[("paging.order", "create")]),
        DEFAULT_LIMIT,
        MAX_LIMIT,
    )
    .unwrap();
    let by_create = apply_paging(&by_create_data, Some(second.created), &create_request);
    assert_eq!(
        by_create
            .resources
            .iter()
            .map(|r| r.id.as_str())
            .collect::<Vec<_>>(),
        [second.id.as_str(), first.id.as_str()],
        "create order should put the most recently created first",
    );
}

#[test]
fn a_zero_limit_returns_no_records_but_still_describes_the_window() {
    // AMWA IS-04-02 `test_21_4`: a 200 with an empty body and
    // `X-Paging-Limit: 0`. The window has no extent, so both cursors sit on
    // the ceiling.
    let data = sample();
    let page = page_for(&refs(&data), &[("paging.limit", "0")], None);

    assert!(page.resources.is_empty());
    assert_eq!(page.limit, 0);
    assert_eq!(page.since.to_string(), "0:20");
    assert_eq!(page.until.to_string(), "0:20");
}

#[test]
fn a_not_truncated_forward_window_echoes_the_requested_ceiling() {
    // The subtle half of Example 5's rule, and what AMWA `test_21_5` checks:
    // when the whole window fitted, the requested ceiling still describes it.
    //
    // Narrowing to the last *matching* record would move the client's `next`
    // cursor backwards, so a later record that did not match the filter would
    // be replayed on the following page.
    let all = sample();
    // A filter selecting a discontiguous, small subset well below the ceiling.
    let matched = vec![resource(5), resource(6)];
    let request = parse_paging(
        lookup(&[("paging.since", "0:1"), ("paging.until", "0:18")]),
        DEFAULT_LIMIT,
        MAX_LIMIT,
    )
    .unwrap();
    let page = apply_paging(&refs(&matched), Some(all[19].updated), &request);

    assert_eq!(cursors(&page), ["0:6", "0:5"]);
    assert_eq!(
        page.until.to_string(),
        "0:18",
        "the ceiling was narrowed although the whole window was served",
    );
}
